Rename storage class into address space

kvark opened the pull request:
Closes #2440
I have no hard feelings about "memory class" as a name, as long as it's not colliding with the other uses of "storage" in the spec. Currently, it's used in 3 different things:
- "storage class" as a term
- "storage" storage class as one of the variants, used for buffers
- "texture_storage" aka the storage texture

Edit: new name is "address space"


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/2524.html" title="Last updated on Jan 25, 2022, 10:15 PM UTC (acd0b72)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2524/483d761...kvark:acd0b72.html" title="Last updated on Jan 25, 2022, 10:15 PM UTC (acd0b72)">Diff</a>

dneto0 commented:
I don't have super strong opinions about this. I haven't read through the change in detail to check for readability.


I used "storage class" initially because 
- it basically lines up with C's storage class concept
- which was then adopted and extended for SPIR-V
- GLSL uses "storage qualifier"

This should go to the W3C group for a decision.

An alternate could be renaming both "storage texture" and "storage buffer" (SSBO).  But I don't have and haven't heard good alternative candidates for those.

litherum commented:
We don't have opinions about this.

dneto0 commented:
Discussed in [2022-01-15 meeting](https://docs.google.com/document/d/1XiSmmqsTfwfw_PNo-pRETKnzd4U1noRPE6uiQYSFRV0/edit#)

Decided to rename "storage class" to "address space"

kvark commented:
Switched to "address space" and rebased. All ready to go.
@dneto0 one thing I'm not sure about is the use of sentences like "memory of a variable", which used to be "storage of a variable". Let me know if you want these to be fixed/adjusted, please.

dneto0 commented on wgsl/index.bs:
```diff
@@ -1085,12 +1085,12 @@ An <dfn noexport>atomic type</dfn> encapsulates a [=scalar=] type such that:
-storage class or by [=storage buffer=] variables with a [=access/read_write=] access mode.
-The [=memory scope=] of operations on the type is determined by the storage
+Atomic types may only be instantiated by variables in the [=address spaces/workgroup=]
+space or by [=storage buffer=] variables with a [=access/read_write=] access mode.
```
"space" -> "address space"

dneto0 commented on spec/index.bs:
```diff
@@ -4933,11 +4933,11 @@ typedef double GPUPipelineConstantValue; // May represent WGSL's bool, f32, i32,
                     <dl class=switch>
                         : {{GPUBufferBindingType/"uniform"}}
-                        :: |variable| is declared with the storage class `uniform`.
+                        :: |variable| is declared with the address space `uniform`.
```
Optional:  Suggest deleting "the" here.

Somehow that reads better?

kvark replied:
I find "the" perfectly appropriate here, as a non-native speaker :)
will do the suggested change anyway

dneto0 commented on wgsl/index.bs:
```diff
@@ -1085,12 +1085,12 @@ An <dfn noexport>atomic type</dfn> encapsulates a [=scalar=] type such that:
-The [=memory scope=] of operations on the type is determined by the storage
+Atomic types may only be instantiated by variables in the [=address spaces/workgroup=]
+space or by [=storage buffer=] variables with a [=access/read_write=] access mode.
+The [=memory scope=] of operations on the type is determined by the memory
```
"memory class" -> "address space"

dneto0 commented on wgsl/index.bs:
```diff
@@ -1085,12 +1085,12 @@ An <dfn noexport>atomic type</dfn> encapsulates a [=scalar=] type such that:
 class it is instantiated in.
-Atomic types in the [=storage classes/workgroup=] storage class have a memory
-scope of `Workgroup`, while those in the [=storage classes/storage=] storage
+Atomic types in the [=address spaces/workgroup=] s[ace] have a memory
```
"s[ace]" -> "address space"

dneto0 commented on wgsl/index.bs:
```diff
@@ -1085,12 +1085,12 @@ An <dfn noexport>atomic type</dfn> encapsulates a [=scalar=] type such that:
-scope of `Workgroup`, while those in the [=storage classes/storage=] storage
+Atomic types in the [=address spaces/workgroup=] s[ace] have a memory
+scope of `Workgroup`, while those in the [=address spaces/storage=]
 class have a memory scope of `QueueFamily`.
```
"class" -> "address space"

dneto0 commented on wgsl/index.bs:
```diff
@@ -1654,7 +1654,7 @@ For a write-only storage texture, the underlying texels are write-only.
 <table class='data'>
   <thead>
-    <tr><th>WGSL storage class<th>SPIR-V storage class
+    <tr><th>WGSL address space<th>SPIR-V address space
```
Keep it "SPIR-V storage class"

kvark replied:
hah, nice catch!

dneto0 commented on wgsl/index.bs:
```diff
@@ -2156,29 +2156,29 @@ The access mode of a memory view must be supported by the storage class. See [[#
 </table>
 
-When *analyzing* a [SHORTNAME] program, reference and pointer types are fully parameterized by a storage class,
+When *analyzing* a [SHORTNAME] program, reference and pointer types are fully parameterized by a address space,
```
"a address space" -> "an address space"

dneto0 commented on wgsl/index.bs:
```diff
@@ -10003,11 +10001,10 @@ between atomic accesses acting on different memory locations.
-classes/workgroup=].
-[=storage classes/workgroup=] atomics have a **Workgroup**
-[=memory scope=] in SPIR-V, while [=storage classes/storage=] atomics have a
+The address space `SC` of the `atomic_ptr` parameter in all atomic built-in
```
I filed a TODO to update the "SC" pattern variables to "AS".

kvark replied:
I hesitated to call it "AS" because there is a different parameter (access mode) called "A".
I wanted to try just "S" (for "space"), so let me know if that doesn't work for you :)

dneto0 commented:
> Switched to "address space" and rebased. All ready to go. @dneto0 one thing I'm not sure about is the use of sentences like "memory of a variable", which used to be "storage of a variable". Let me know if you want these to be fixed/adjusted, please.

Those uses look good to me. Thanks for highlighting them.

dneto0 reviewed (changes requested):
I found some missing instances, and minor wording fixes.

kvark commented:
Addressed everything and rebased. Thank you for the thorough review!

dneto0 reviewed (approved):
Thanks!

kdashg commented:
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-01-25-minutes#rename-storage-class-into-memory-class-2524">WGSL meeting minutes 2022-01-25</a></summary>



* DM: lots of terms use storage. Too many IMO. I picked changing storage class to memory class. Don’t think David’s reasoning is that strong
* DM: “memory class” is fine with me as standalone, detached from past cultural reference.
* AB: This is a new term of art.  Would like reuse a name like “address space” used elsewhere. “Memory class” doesn’t conjure something specific to me.
* AB: HLSL’s “constant” buffer  for UBO
* JB: MSL uses “address space”.
* AB: OpenCL uses address space, lots of compilers use it.
* MM: In Metal it is a different address space.
* JB: We segregate these because they have different implementations.
* _Consensus: rename “storage class” to “address space”.  DM to pick it up._


</details>
